### DISCUSSION TITLE
Can we skip some fixed pattern stack backtrace in cstack=vm ?

### DISCUSSION THREAD

**Discussion by byte0100** [async-profiler/async-profiler#995] [Q&A] [answered]
Compare to AGCT, StackWalker::walkVM() would always output fixed pattern  of stack frames.
e.g. in stack bottom, like this:
...
frist_java_frame;
.call_stub()L;
libjvm.so.JavaCalls::call_helper()L;
libjvm.so.JavaCalls::call_virtual()L;
libjvm.so.thread_entry()L;
libjvm.so.JavaThread::run()L;
libjvm.so.Thread::call_run()L;
libjvm.so.thread_native_entry()L;
libpthread-2.24.so.start_thread()L;
libc-2.24.so.__clone()L;

When calling the java Reflection method（Method#invoke:

.call_stub()L;
libjvm.so.JavaCalls::call_helper()L;
libjvm.so.invoke()L;
libjvm.so.Reflection::invoke_method()L;
libjvm.so.JVM_InvokeMethod()L;

These frames are the fixed pattern. Can we just unwind these frames but not to fill the frames, like the AGCT stack backtrace results.

**Comment by apangin** [comment, marked as answer]
With `cstack=vm`, async-profiler collects full stack trace of a thread, including all intermediate JVM frames. This is intentional. It was exactly the reason why `vm` mode appeared - completeness was one of the main requirements from customers (see #795).

If you are not interested in such level of detail, use other stack walking modes (`cstack=dwarf` may be a good option). Alternatively, you may post-process recordings to exclude frames you don't need.

**Reply by byte0100** [reply]
Thank you for your prompt reply. 
We want to use  the vm mode to do stack backtrace of the Java frames , for performance reasons, instead of AGCT. So we want to get rid of these frames that aren't interesting.Can you provide some suggestions for achieving this requirement?
Thanks again for the help!

**Reply by apangin** [reply]
`cstack=vm` is not more performant than AsyncGetCallTrace. If you are not interested in intermediate frames, `cstack=dwarf` should be sufficient.

**Comment by richardstartin** [comment]
@apangin since you have the start and end addresses for the call stub for other reasons, what do you think about adding an option to terminate the unwind when hitting the call stub? In my experience, Java users don’t care about how the VM runs the Java program, and should focus on things they can control, i.e. most frames below the call stub. If this information is useful to specialist users, the option to unwind further should be available.

**Reply by apangin** [reply]
This is what the default stack walking mode does, isn't it?

**Reply by richardstartin** [reply]
I thought the primary motivation for the VMStructs unwinder was not crashing in AsyncGetCallTrace? If the VMStructs unwinder is safer, it would be nice if it could be configured to skip frames above the call stub that many (most?) users will find irrelevant.

**Reply by apangin** [reply]
Motivation for this change is discussed in #795: reliability was one of the reasons, but not the only one.
AsyncGetCallTrace is also supposed to be stable. If it crashes, it is a JVM bug that needs to be fixed.
`cstack=vm` has always been considered a mode for advanced users. That said, I admit there are use cases for a "light" version of VMStructs stack walker and I'm happy to review a pull request if it does not complicate things too much.

**Reply by apangin** [reply]
"Light" mode of VMStructs stack walker has been added in #1073.
